draft-normington-jsonpath-latest.md

```diff
@@ -84,43 +108,266 @@ informative:
 
 normative:
   RFC3629:
-  RFC5234:
+  RFC5234: abnf
+  RFC8259: json
 
 --- abstract
 
-
 JSONPath defines a string syntax for identifying values
 within a JavaScript Object Notation (JSON) document.
 
 --- note_
 
 **This document is a work in progress and has not yet been published
-as an Internet Draft.**
+as an Internet Draft** (which needs to be fixed soon).
+
+--- note_Contributing
+
+This document picks up the popular JSONPath specification dated
+2007-02-21 and provides a more normative definition for it.
+In its current state, it is a strawman document showing what needs to
+be covered.
+
+Comments and issues can be directed at the github repository *insert
+repo here* as well as (for the time when the more permanent home is
+being decided) at the dispatch@ietf.org mailing list.
 
 --- middle
 
 # Introduction
 
-JSONPath was introduced by Stefan Goessner as a simple
-form of XPath for JSON.
-See his original article {{Goessner}}.
+This document picks up the popular JSONPath specification dated
+2007-02-21 {{JSONPath-orig}} and provides a more normative definition for it.
+In its current state, it is a strawman document showing what needs to
+be covered.
 
 JSON is defined by {{RFC8259}}.
 
-## Requirements Language
+JSONPath is not intended as a replacement, but as a more powerful
+companion, to JSON Pointer {{RFC6901}}. \[insert reference to section
+where the relationship is detailed.]
 
-{::boilerplate bcp14}
 
-## ABNF Syntax
+## Terminology
 
-The syntax in this document conforms to ABNF as defined by {{RFC5234}}.
+{::boilerplate bcp14}
 
+The grammatical rules in this document are to be interpreted as ABNF,
+as described in {{-abnf}}.
 ABNF terminal values in this document define Unicode code points rather than
 their UTF-8 encoding.
 For example, the Unicode PLACE OF INTEREST SIGN (U+2318) would be defined
 in ABNF as `%x2318`.
 
 
+The terminology of {{-json}} applies.
+
+Data Item:
+: A structure complying to the generic data model of JSON, i.e.,
+  composed of containers such as arrays and maps (JSON objects), and
+  of atomic data such as null, true, false, numbers, and text strings.
+
+Object:
+: Used in its generic sense, e.g., for programming language objects.
+  When a JSON Object as defined in {{-json}} is meant, we specifically
+  say JSON Object.
+
+Query:
+: Short name for JSONPath expression.
+
+Argument:
+: Short name for the JSON data item a JSONPath expression is applied to.
+
+Output Path:
+: A simple form of JSONPath expression that identifies a Position by
+  providing a query that results in exactly that position.  Similar
+  to, but syntactically different from, a JSON Pointer {{-pointer}}.
```

Comment:
We could.  There are subtle differences we'd need to elaborate first to do this.

Reply:
I don't think you can use JSONPointer for that.  JSONPointer only allows for locating primitive values, it doesn't work for empty arrays or empty objects.

Reply:
But can a JSONPointer identify an empty object or empty array? My understanding of https://tools.ietf.org/html/rfc6901#section-7 is that that is not defined.

Reply:
Sure.  For `{ "a": {} }`, `/a` points to the empty object.  What in Section 7 makes you think it doesn't?

Reply:
Specifically, JSON Schema uses JSON Pointers in its `$ref` keyword to reference a location within the schema in its JSON form.  Per JSON Schema, the reference MUST resolve to a valid schema object.  However, `{}` is a perfectly valid object.

The only time Section 7 applies is when the pointer indicates a route _into_ this object, e.g. `/a/missing` in @cabo's example would be required to error because the value at `/a` doesn't contain a `missing` property.

Reply:
RFC 6901 goes to lengths to make sure any JSON map key (which essentially is any Unicode string) can be used as a path component.  It decodes that (needing an escape for / and then for the escape), and then looks whether it is looking at an array or a map.  Any maps is directly indexed, while for an array the path component is interpreted as a decimal number (there also is the "-" special, which I think we don't need).  The latter is easy.  Doing the former requires having a bracket/string syntax for any Unicode string.

Reply:
Maybe just something to consider for the future, but not relevant for this PR, then.

Reply:
You're right, thanks for the correction.

Reply:
I've opened #44 for this.